src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,207 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> prev;
+        private Node<T> next;
+
+        Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index);
+
+        if (size == 0) {
+            addFirstElementToList(value);
+            return;
+        }
+
+        if (index == 0) {
+            addFirst(value);
+            return;
+        }
+        if (index == size) {
+            addLast(value);
+            return;
+        }
+
+        insertNode(value, index);
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
+        checkIndexAdvanced(index);
+
+        Node<T> node = head;
+        for (int i = 0; i < size; i++) {
+            if (i == index) {
+                return node.value;
+            }
+            node = node.next;
+        }
         return null;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexAdvanced(index);
+
+        Node<T> node = head;
+        T oldValue = null;
+        for (int i = 0; i < size; i++) {
+            if (i == index) {
+                oldValue = node.value;
+                node.value = value;
+            }
+            node = node.next;
+        }
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexAdvanced(index);
+        if (size == 1) {
+            head = tail = null;
+            size--;
+            return null;
+        }
+        if (index == 0) {
+            return removeFirst();
+        }
+        if (index == size - 1) {
+            return removeLast();
+        }
+
+        Node<T> node = head;
+        T removedValue = null;
+        for (int i = 0; i < size; i++) {
+            if (i == index) {
+                removedValue = node.value;
+                node.prev.next = node.next;
+                node.next.prev = node.prev;
+                size--;
+                return removedValue;
+            }
+            node = node.next;
```

Comment:
You should create a separate method unlink that will take a Node that needs to be removed and unlink it. 

https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html#create-a-separate-method-unlinknode-node